components/style/values/specified/image.rs

```diff
@@ -0,0 +1,370 @@
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ColorStop {
+    /// The color of this stop.
+    pub color: CSSColor,
+
+    /// The position of this stop. If not specified, this stop is placed halfway between the
+    /// point that precedes it and the point that follows it.
+    pub position: Option<LengthOrPercentage>,
+}
+
+impl ToCss for ColorStop {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        try!(self.color.to_css(dest));
+        if let Some(position) = self.position {
+            try!(dest.write_str(" "));
+            try!(position.to_css(dest));
+        }
+        Ok(())
+    }
+}
+
+define_css_keyword_enum!(HorizontalDirection: "left" => Left, "right" => Right);
+define_css_keyword_enum!(VerticalDirection: "top" => Top, "bottom" => Bottom);
+
+fn parse_one_color_stop(input: &mut Parser) -> Result<ColorStop, ()> {
+    Ok(ColorStop {
+        color: try!(CSSColor::parse(input)),
+        position: input.try(LengthOrPercentage::parse).ok(),
+    })
+}
+
+/// Determines whether the gradient's ending shape is a circle or an ellipse.
+/// If <shape> is omitted, the ending shape defaults to a circle
+/// if the <size> is a single <length>, and to an ellipse otherwise.
+#[derive(Clone, PartialEq, Debug)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum EndingShape {
+    Circle(Size<Length>),
+    Ellipse(Size<LengthOrPercentage>, Size<LengthOrPercentage>),
+}
+
+impl Parse for EndingShape {
+    fn parse(input: &mut Parser) -> Result<Self, ()> {
+        match_ignore_ascii_case! { try!(input.expect_ident()),
+            "circle" => {
+                let position = input.try(Size::parse).unwrap_or(
+                    Size::Keyword(SizeKeyword::FarthestSide));
+                Ok(EndingShape::Circle(position))
+            },
+            "ellipse" => {
+              let (first_pos, second_pos) = if let Ok(first) = input.try(Size::parse) {
+                    if let Ok(second) = input.try(Size::parse) {
+                        (first, second)
+                    } else {
+                        (first, Size::Keyword(SizeKeyword::FarthestSide))
+                    }
+                } else {
+                    (Size::Keyword(SizeKeyword::FarthestSide),
+                     Size::Keyword(SizeKeyword::FarthestSide))
+                };
+                Ok(EndingShape::Ellipse(first_pos, second_pos))
+            },
+            _ => {
+                // If 1 <length> is present, it defaults to circle, otherwise defaults to ellipse.
+                if let Ok(first) = input.try(Size::parse) {
+                    if let Ok(second) = input.try(Size::parse) {
+                        Ok(EndingShape::Ellipse(first, second))
+                    } else {
+                        if let Size::Length(length) = first {
+                            if let LengthOrPercentage::Length(len) = length {
```

Comment:
It can also be calc. But the calc in `Length` parses differently from the one in `LengthOrPercentage` (`CalcLoP::parse_length()` vs `CalcLoP::parse_length_or_percentage()`). So we can't use the result of a parsed LoP here.

Instead do something like this:

```rust
if let Ok((first,second)) = input.try(|input| { let first = try!(Size<LoP>::parse); let second = try!(Size<LoP>::parse); (first, second)}) {
 // handle ellipse
} else {
 // handle circle
}
```

(might want to indent it better or move the closure to a previous line)